pkg/tools/fifosync/mutex.go

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fifosync provides FIFO order guaranteed synchronization primitives
+package fifosync
+
+import (
+	"sync"
+)
+
+// Mutex is a FIFO order guaranteed mutex
+type Mutex struct {
+	init sync.Once
+	ch   chan struct{}
+}
+
+// Lock acquires lock or blocks until it gets free
+func (m *Mutex) Lock() {
+	m.init.Do(func() {
+		m.ch = make(chan struct{}, 1)
+	})
+	m.ch <- struct{}{}
+}
+
+// Unlock frees lock
+func (m *Mutex) Unlock() {
+	<-m.ch
+}
+
+// Mutate mutates 'locked' Mutex to 'next' Mutex:
+// * if 'next' mutex is free, it first locks 'next', than unlocks 'locked'
+// * if 'next' mutex is locked, it first unlocks 'locked', than locks (or starts
+//   waiting on) 'next'
+func Mutate(locked, next *Mutex) {
+	next.init.Do(func() {
+		next.ch = make(chan struct{}, 1)
```

Comment:
This doesn't preserve order.  Try running:

```go
func BenchmarkMutex_Lock_Different_Goroutine_Order(b *testing.B) {
	var g fifosync.Mutex
	check := map[int]bool{0:true}
	for i := 0; i<b.N; i++ {
		go func(i int) {
			g.Lock()
			if !check[i]{
				b.Errorf("did not find expected value in check[%d]",i)
			}
			check[i+1] = true
			g.Unlock()
		}(i)
	}
}
```

Reply:
This test is not actually correct, because such code:
```go
go funcA()
go funcB()
go funcC()
```
doesn't mean that the order of execution will be `funcA()`, `funcB()`, `funcC()`.
For the order preserving testing I have added `<-time.After(time.Millisecond)` in my tests to let the goroutines start:
https://github.com/networkservicemesh/sdk/blob/a84477b8e856e9799a81745d91489f6bb5e15dfa/pkg/tools/fifosync/mutex_test.go#L36-L50

Reply:
Change your time.Sleep to microsecond and see what happens.

What you've really demonstrated here is why Mutex semantics are not appropriate for preserving order.  The issue isn't one of implementation, its one of *semantics*.  The analogous test for AsyncExec:

```
func BenchmarkAsyncExec_Order(b *testing.B) {
	var g fifosync.Map
	check := map[int]bool{0:true}
	for i := 0; i<b.N; i++ {
		j := i
		g.AsyncExec("foo",func(){
			if !check[j]{
				b.Errorf("did not find expected value in check[%d]",i)
			}
			check[j+1] = true
		})
	}
}
```

Passes just fine, no sleep hacking needed to make it work.  I *does* guarantee order.  go routines + mutexes never do, and adding sleeps to try to make them isn't really the answer.

Reply:
> What you've really demonstrated here is why Mutex semantics are not appropriate for preserving order. The issue isn't one of implementation, its one of semantics. The analogous test for AsyncExec passes just fine, no sleep hacking needed to make it work. I does guarantee order. go routines + mutexes never do, and adding sleeps to try to make them isn't really the answer.

I guess that the use case we want to cover with `MutexGroup` / `Map` is **preserving order of execution** for multiple events coming from **multiple goroutines** in some **absolute order**.
Your test is about the **execution order preserving** for the events coming from the **single goroutine** and it does assume some **absolute order** because single threaded execution is serial by nature. But it tells nothing about the **multiple goroutines** case.
Such sleep / waitgroup hacking does nothing with the **order of execution** because [in my tests the mutex is initially locked](https://github.com/networkservicemesh/sdk/blob/5df451b681f28bbc9997ce260b1a5e8d85ef59a8/pkg/tools/fifosync/mutex_group_test.go#L32). It only presumes the **order of locking**. We need it, because there is no other way to guarantee the **absolute order** of the incoming events in the **multiple goroutines** case.

> Change your time.Sleep to microsecond and see what happens.

Such thing breaks the **order of locking** so the **absolute order** of the incoming events becomes different and so it does mean that the **order of execution** must become different.